src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt

```diff
@@ -0,0 +1,47 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.ApplicationListener
+import org.springframework.stereotype.Component
+import org.springframework.web.util.UriComponentsBuilder
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.NotifyService
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.SNSService
+
+@Component
+class ReferralEventListener(
+  private val notifyService: NotifyService,
+  private val snsService: SNSService,
+  private val objectMapper: ObjectMapper,
+  @Value("\${interventions-ui.baseurl}") private val interventionsUIBaseURL: String,
+  @Value("\${interventions-ui.locations.sent-referral}") private val interventionsUISentReferralLocation: String,
+) : ApplicationListener<ReferralEvent> {
+  override fun onApplicationEvent(event: ReferralEvent) {
+    when (event.type) {
+      ReferralEventType.SENT -> {
+        sendReferralSentEmail(event)
+        publishReferralSentToSNSTopic(event)
+      }
+    }
+  }
+
+  private fun sendReferralSentEmail(event: ReferralEvent) {
+    val location = UriComponentsBuilder.fromHttpUrl(interventionsUIBaseURL)
+      .path(interventionsUISentReferralLocation)
+      .buildAndExpand(event.referral.id)
+      .toUri()
+
+    // fixme: this email address will eventually come from the provider associated with the referral
+    notifyService.referralSent(event.referral.referenceNumber!!, location, "tom.myers@digital.justice.gov.uk")
+  }
+
+  private fun publishReferralSentToSNSTopic(event: ReferralEvent) {
+    // fixme: what data needs to go in here
+    val referralSentMessage = mapOf(
+      "type" to "referral-sent",
+      "description" to "A referral has been sent to a Service Provider",
+      "id" to event.referral.id,
+    )
+    snsService.publish(objectMapper.writeValueAsString(referralSentMessage))
```

Comment:
I think the `fixme` comment highlights that this class might prove difficult to test:

- there's logic about location URI building for emails that need testing
- there's logic about constructing a referral SNS message that needs testing

Those tests don't seem related.

Reply:
agreed - changed.